[HUDI-3742] Enable parquet enableVectorizedReader for spark inc query to improve peformance

xiarixiaoyao opened the pull request:
### Change Logs
add config to enable parquet enableVectorizedReader for spark inc query
### Impact
improve incremental query for spark.

### Risk level (write none, low medium or high below)
low

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

XuQianJin-Stars commented:
LGTM

xiarixiaoyao commented:
@xushiyan @alexeykudinkin @nsivabalan 
pls help me check this pr, thanks
for incremental read parquet file, we can achive 2.x faster

alexeykudinkin commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:
```diff
@@ -59,6 +62,16 @@ class MergeOnReadIncrementalRelation(sqlContext: SQLContext,
     super.timeline.findInstantsInRange(startTimestamp, endTimestamp)
   }
 
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
```
@xiarixiaoyao we don't need to override `buildScan` for that you can do the same thing w/in `composeRDD`

xiarixiaoyao replied:
ok， let me fixed it

alexeykudinkin commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:
```diff
@@ -95,6 +108,9 @@ class MergeOnReadIncrementalRelation(sqlContext: SQLContext,
     )
 
+    // setUp tableRequiredSchema
+    tableRequiredSchema = requiredSchema.structTypeSchema
```
Why do we need this one? Can we use `requiredSchema` directly?

xiarixiaoyao replied:
> @xiarixiaoyao while i understand your intent that we can speed things up for some of the queries, i don't think enabling vectorized reader for Incremental queries will bring an universal speed up for all queries -- vectorized reader would mean that we first fetch records and only do filtering in memory, which i could clearly could be a disadvantage for large tables.
> 
> So my suggestion would be to NOT enable vectorized-reader for Incremental queries by default, but instead let individual users to decide on whether they want to enable it or not.

@alexeykudinkin  I don't think record level filter has better performance. Spark is off by default for this parameter. 
Read out the data, vectorize it, and then filter it better. I will take out 100g data to give the test results

alexeykudinkin replied:
@xiarixiaoyao sorry for the late reply.

I understand that it reads it by batches, but my point still stands -- unfortunately, (i don't really understand why those 2 things are incompatible, but nevertheless) when we use vectorization -- Parquet filter push-down stops working, which means that now every row-group would be read in memory, before being filtered out as irrelevant. 

For use-cases with localized storage (ie adjacent to the Compute) it might not be a problem, but it's more likely to be a problem for use-cases with disaggregated Compute and Storage (for ex, in Cloud) -- now we will have to read whole row-group for Spark to be able to filter it out in memory.

I'd suggest additionally to the Benchmark that you ran locally, can we also test similar setup in Cloud (like AWS, or wherever), so that we get a better sense what the impact would be in Cloud env?

alexeykudinkin reviewed (commented):
@xiarixiaoyao while i understand your intent that we can speed things up for some of the queries, i don't think enabling vectorized reader for Incremental queries will bring an universal speed up for all queries -- vectorized reader would mean that we first fetch records and only do filtering in memory, which i could clearly could be a disadvantage for large tables.

So my suggestion would be to NOT enable vectorized-reader for Incremental queries by default, but instead let individual users to decide on whether they want to enable it or not.

garyli1019 commented:
IIRC the vectorized reader was disabled for a reason. somehow the filtering was depends on the regular parquet reader.

xiarixiaoyao commented:
> IIRC the vectorized reader was disabled for a reason. somehow the filtering was depends on the regular parquet reader.

@garyli1019  thanks for your comments.
yes， i know the reason。so in this pr, I directly generated the code for the filter, vectorized the data, and directly filtered it after reading it

vinothchandar commented:
@xiarixiaoyao can you post any test results for this effort.

vinothchandar commented:
@xiarixiaoyao Main contention here is to introduce a config and keep existing behavior

xiarixiaoyao commented:
> @xiarixiaoyao can you post any test results for this effort.

Sorry for the late reply， let me do it now

xiarixiaoyao commented:
@vinothchandar  @alexeykudinkin @garyli1019 
addressed all comments
add config to keep original logical.
benchmark setUp
1) orignal table: 1000w row,  10 columns,   
2) insert 100w 
3) query begin incremental from commit time 0
benchmark result:
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_201-b09 on Windows 10 10.0
Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
perf mor incr read:                       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
vectorized disable                                 1730           1796          94          5.8         173.0       1.0X
vectorized enable                                   739            769          26         13.5          73.9       2.3X
```

do we need add benchmark code to this pr, if needed, i will added

alexeykudinkin commented:
@xiarixiaoyao can you please rebase this one? 

Also please fill out the description to explain the context.

xiarixiaoyao commented:
@alexeykudinkin sorry to miss your message。 will update the code and add benmark result for it

nsivabalan commented:
@xiarixiaoyao : gentle reminder.

xiarixiaoyao commented:
@hudi-bot run azure

xiarixiaoyao commented:
@hudi-bot run azure

xiarixiaoyao commented:
@alexeykudinkin @nsivabalan 
rebase the codes，
Add parameters to control this behavior, which is enabled by default

xiarixiaoyao commented:
@hudi-bot run azure

alexeykudinkin commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/DataSourceOptions.scala:
```diff
@@ -152,6 +152,11 @@ object DataSourceReadOptions {
+  val INCREMENTAL_VECTORIZED_READER_ENABLE: ConfigProperty[String] = ConfigProperty
+    .key("hoodie.datasource.read.incr.vectorized.reader.enable")
+    .defaultValue("true")
+    .withDocumentation("Enable vectorized reader for mor table incremental query.")
```
This config should be applicable for both COW/MOR

alexeykudinkin commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieBaseRelation.scala:
```diff
@@ -347,14 +347,25 @@ abstract class HoodieBaseRelation(val sqlContext: SQLContext,
+    * @param requiredSchema  projected schema required by the reader
+    * @return handled RDD
+    */
+  protected def postOperationOnComposeRDD(rdd: RDD[InternalRow], requiredSchema: HoodieTableSchema): RDD[Row] = {
```
Can we implement this custom logic w/in the `composeRDD` itself? Why do we need a separate method for it?

alexeykudinkin commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:
```diff
@@ -122,6 +144,17 @@ class MergeOnReadIncrementalRelation(sqlContext: SQLContext,
+    val filterExpression = convertToExpressions(incrementalSpanRecordFilters.toArray).map { e =>
+      e transform {
+        case a: AttributeReference =>
+          attrs.find(_.name.equalsIgnoreCase(a.name)).get
```
Instead of manipulating expressions like that we should instead fix `convertToExpressions` by:
 - Making it produce unresolved expressions, replacing name-refs w/ `UnresolvedAttribute` instead of `AttributeReference` during conversion
 - Then resolve converted expression using `resolveExpr` (in the same utility)

xiarixiaoyao commented:
> IIRC the vectorized reader was disabled for a reason. somehow the filtering was depends on the regular parquet reader.

garyli1019  has explained the problem: IIRC the vectorized reader was disabled for a reason. somehow the filtering was depends on the regular parquet reader.

Personally,i think we should enable vectorization by default

YannByron commented:
> Personally,i think we should enable vectorization by default

+1. Users do not know when to enable vectorization. So provide a config to control this is meaningless.

bvaradar commented:
@xiarixiaoyao : Can you address the comments in the PR ? 
@garyli1019 : Any other concern about having vectorization for incr query for MOR  (with default turned off ? )

garyli1019 commented:
@bvaradar only partial columns will return by the vectorized reader, but the merging with log might need some columns not in the select statement, or the necessary columns using by hudi, such as precombine field or primary key. How do we handle this case?

bvaradar commented:
Good point.  It doesn't look like the place where we set spark vectorized reading on has knowledge of read schema to safely enable vectorization for certain cases for MOR. @xiarixiaoyao : Thoughts ?

xiarixiaoyao commented:
close this pr as we no longer need it
